### PR TITLE
fix(mcp): count runes, not bytes, when flagging mem_search previews as truncated

### LINKED ISSUE #751: bug(mcp): the [preview] marker counts bytes while the cut counts runes
### 📋 Pre-flight Checks

- [x] I have searched existing issues and this is not a duplicate
- [x] I understand this issue needs `status:approved` before a PR can be opened

### 📝 Bug Description

- `truncate()` cuts previews at 300 runes, but the `[preview]` marker and the "Results above are previews (300 chars)" footer trigger when content exceeds 300 BYTES.
- Non-ASCII content (accents, CJK, emoji) between 301 bytes and 300 runes is shown in full yet labeled `[preview]`.
- That label tells the agent to spend a `mem_get_observation` call on content it already has.

### 🔄 Steps to Reproduce

1. Save an observation whose content is 250 accented characters (over 300 bytes, under 300 runes).
2. Run `mem_search` and find it.
3. The full content is shown with `[preview]` appended, plus the preview footer.

### ✅ Expected Behavior

The marker appears only when the preview is actually truncated.

### ❌ Actual Behavior

The marker appears on content that is already complete.

### Operating System

macOS

### Engram Version

1.20.0 (`internal/mcp/mcp.go`, the `len(r.Content) > 300` check)

### Agent / Client

MCP (any agent)

### 💡 Additional Context

- One-line fix: `len([]rune(r.Content)) > 300`, matching the rune-based cut `truncate()` already does.
- Fix branch ready with a regression test. Will open the PR once this gets `status:approved`.

### PR DESCRIPTION
**Pull request by yudelsuarez**
## 🔗 Linked Issue

Closes #751

---

## 🏷️ PR Type

- [x] `type:bug` — Bug fix
- [ ] `type:feature` — New feature
- [ ] `type:docs` — Documentation only
- [ ] `type:refactor` — Code refactoring (no behavior change)
- [ ] `type:chore` — Maintenance, dependencies, tooling
- [ ] `type:breaking-change` — Breaking change

---

## 📝 Summary

- `handleSearch` builds each preview with `truncate(r.Content, 300)`, which slices **runes**, but decided whether to append the `[preview]` marker and the previews footer from `len(r.Content)`, which counts **bytes**.
- Content between 301 bytes and 300 runes was therefore emitted **whole** and still labelled as truncated, so the agent spends a `mem_get_observation` call on content it already has in full. Accented Latin text, CJK and emoji cross that gap routinely.
- The comparison now counts the same unit `truncate` cuts on, using `utf8.RuneCountInString`, the form already used elsewhere in this repo. What gets cut, and where, is untouched.

## 📂 Changes

| File | Change |
|------|--------|
| `internal/mcp/mcp.go` | Marker and footer decision only: `len(r.Content) > 300` becomes `utf8.RuneCountInString(r.Content) > 300`, plus the `unicode/utf8` import |
| `internal/mcp/mcp_test.go` | `TestHandleSearch_PreviewMarkerCountsRunes` pins the boundary from both sides and on the limit itself |

## 🧪 Test Plan

- [x] Unit tests pass locally: `go test ./...`
- [x] E2E tests pass locally: `go test -tags e2e ./internal/server/...`
- [x] Plugin tests pass locally: `npm test` in `plugin/pi`, 81 passed
- [x] Manually tested the affected functionality

What was run, and what came out:

- `go test ./internal/mcp/...` green, including the new test.
- `go test -tags e2e ./internal/server/...` green.
- `npm test` in `plugin/pi`: 81 passed, 0 failed.
- A full `go test ./...` on macOS also surfaces one failure in `cmd/engram` that is unrelated to this change and reproduces identically on an unmodified `main`, so it is not introduced here.
- Manual check: saved an observation of 250 accented characters and called `mem_search`. The result came back whole, with no `[preview]` marker and no footer. Before the change the same observation came back whole and marked.

The test brackets the boundary so it cannot pass for the wrong reason:

| Case | Runes | Bytes | Cut by `truncate`? | Marker expected |
|------|-------|-------|--------------------|-----------------|
| whole content is never a preview | 200 | 400 | no | no |
| content exactly at the limit is not cut | 300 | 600 | no | no |
| cut content is still a preview | 301 | 602 | yes | yes |

Every case uses accented runes, so byte length always exceeds rune length. The last case is what stops the change from over-correcting: genuinely truncated content must still be flagged.

Reverting the one-line change fails the first two cases, on both the marker and the footer, and leaves the third green:

```
--- FAIL: TestHandleSearch_PreviewMarkerCountsRunes (0.15s)
    --- FAIL: TestHandleSearch_PreviewMarkerCountsRunes/whole_content_is_never_a_preview (0.05s)
        mcp_test.go:6572: [preview] marker present = true, want false for 200 runes (400 bytes)
        mcp_test.go:6576: preview footer present = true, want false for 200 runes (400 bytes)
    --- FAIL: TestHandleSearch_PreviewMarkerCountsRunes/content_exactly_at_the_limit_is_not_cut (0.05s)
        mcp_test.go:6572: [preview] marker present = true, want false for 300 runes (600 bytes)
        mcp_test.go:6576: preview footer present = true, want false for 300 runes (600 bytes)
FAIL
FAIL	github.com/Gentleman-Programming/engram/internal/mcp	0.408s
```

Each assertion also prints the full search response for context; those lines are elided above.

---

## 🤖 Automated Checks

These run automatically and **all must pass** before merge:

| Check | What it verifies | Status |
|-------|-----------------|--------|
| **Check Issue Reference** | PR body contains `Closes #N` / `Fixes #N` / `Resolves #N` | ⏳ |
| **Check Issue Has status:approved** | Linked issue has `status:approved` label | ⏳ |
| **Check PR Has type:\* Label** | PR has exactly one `type:*` label | ⏳ |
| **Unit Tests** | `go test ./...` passes | ⏳ |
| **E2E Tests** | `go test -tags e2e ./internal/server/...` passes | ⏳ |
| **Plugin Tests** | `npm test` passes in `plugin/pi` | ⏳ |

---

## ✅ Contributor Checklist

- [x] I linked an approved issue above (`Closes #751`)
- [ ] I added exactly **one** `type:*` label to this PR. Labelling needs write access, so this one is for a maintainer; `type:bug` is the right one
- [x] I ran unit tests locally: `go test ./...`
- [x] I ran e2e tests locally: `go test -tags e2e ./internal/server/...`
- [x] Docs updated (if behavior changed): no documented behaviour changes, nothing in the docs describes the marker
- [x] Commits follow [conventional commits](https://www.conventionalcommits.org/) format
- [x] No `Co-Authored-By` trailers in commits

---

## 💬 Notes for Reviewers

Scope is deliberately narrow. `truncate` was already correct, and the other five `truncate` call sites in `internal/mcp/mcp.go` (lines 1619, 1797, 1807, 1814 and 1821) never pair it with a length comparison, so I found no other place where the two units disagree.

The diff is 73 changed lines in one work unit, inside the 400-line review budget.

## Summary by CodeRabbit

* **Bug Fixes**
  * Search previews now measure content by characters rather than bytes, improving handling of accented and other multibyte text.
  * Preview markers and notices now appear only when content exceeds the 300-character limit.
  * Text containing multibyte characters is no longer truncated prematurely when it remains within the preview limit.

### CODE REVIEW

**Comment by yudelsuarez** [comment]
Added the missing boundary case the review flagged: content of exactly 300 runes (600 bytes) is not cut by `truncate`, so it must not carry the marker either. The test now covers 200, 300 and 301 runes, and reverting the one line fails the first two while leaving the third green.

File: internal/mcp/mcp.go [hunk 1]
```diff
@@ -1045,7 +1045,7 @@ func handleSearch(s *store.Store, cfg MCPConfig, activity *SessionActivity) serv
 				projectDisplay = fmt.Sprintf(" | project: %s", *r.Project)
 			}
 			preview := truncate(r.Content, 300)
-			if len(r.Content) > 300 {
+			if len([]rune(r.Content)) > 300 {
 				anyTruncated = true
 				preview += " [preview]"
 			}
```
[the remark this answers is not in the document]
Comment (yudelsuarez) [reply internal/mcp/mcp.go:1115]: There is no second allocation to remove. The compiler rewrites `len([]rune(s))` into a `runtime.countrunes` call, so the slice is never built:

```
$ go build -gcflags=-S
CALL runtime.countrunes(SB)
```

Benchmarked on a 2220-byte, 2072-rune string, both forms report `0 B/op, 0 allocs/op` and land within 2% of each other.

I switched to `utf8.RuneCountInString` anyway, for a different reason: it is the form already used in this repo (`internal/cloud/dashboard/helpers.go` and `internal/mcp/serverinstructions_length_test.go`), while `len([]rune(...))` appears nowhere else.

I did not take the `preview != r.Content` suggestion. It is correct today, but it infers truncation from `truncate`'s return value instead of stating the condition, so any future change to `truncate` would silently move the marker with it.

File: internal/mcp/mcp.go [hunk 2]
```diff
@@ -1045,7 +1046,7 @@ func handleSearch(s *store.Store, cfg MCPConfig, activity *SessionActivity) serv
 				projectDisplay = fmt.Sprintf(" | project: %s", *r.Project)
 			}
 			preview := truncate(r.Content, 300)
-			if len(r.Content) > 300 {
+			if utf8.RuneCountInString(r.Content) > 300 {
 				anyTruncated = true
 				preview += " [preview]"
```
[the remark this answers is not in the document]
Comment (yudelsuarez) [reply internal/mcp/mcp.go:1114]: Good catch, that was stale. The description now matches the code: both the summary line and the Changes table read `utf8.RuneCountInString(r.Content) > 300`.

**Comment by dnlrsls** [comment]
well done @yudelsuarez

**Comment by dnlrsls** [comment]
@coderabbitai resume
